File: client/java-client/src/main/java/com/kotak/neo/client/NeoAPI.java
package com.kotak.neo.client;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.kotak.neo.client.api.LoginApi;
import com.kotak.neo.client.api.MarginApi;
import com.kotak.neo.client.api.OrderApi;
import com.kotak.neo.client.api.PortfolioApi;
import com.kotak.neo.client.api.QuotesApi;
import com.kotak.neo.client.api.ScripApi;
import com.kotak.neo.client.rest.RestClient;
import com.kotak.neo.client.websocket.Instrument;
import com.kotak.neo.client.websocket.NeoWebSocket;

import java.util.List;
import java.util.function.Consumer;

/**
 * NeoAPI — top-level SDK entry point. Mirrors the Python NeoAPI class.
 */
public class NeoAPI {
    public final Configuration configuration;
    private final RestClient rest;
    private final LoginApi loginApi;
    private final OrderApi orderApi;
    private final PortfolioApi portfolioApi;
    private final MarginApi marginApi;
    private final ScripApi scripApi;
    private final QuotesApi quotesApi;
    private NeoWebSocket neoWebSocket;

    public Consumer<String> onOpen;
    public Consumer<Object> onMessage;
    public Consumer<Throwable> onError;
    public Consumer<String> onClose;

    public NeoAPI() {
        this("uat", null, null, null);
    }

    public NeoAPI(String environment, String accessToken, String neoFinKey, String consumerKey) {
        this.configuration = new Configuration(environment);
        this.configuration.bearerToken = accessToken;
        this.configuration.neoFinKey = neoFinKey;
        this.configuration.consumerKey = consumerKey;
        this.rest = new RestClient(this.configuration);
        this.loginApi = new LoginApi(rest, configuration);
        this.orderApi = new OrderApi(rest, configuration);
        this.portfolioApi = new PortfolioApi(rest, configuration);
        this.marginApi = new MarginApi(rest, configuration);
        this.scripApi = new ScripApi(rest, configuration);
        this.quotesApi = new QuotesApi(rest, configuration);
    }

    private JsonObject notLoggedInError() {
        JsonObject err = new JsonObject();
        err.addProperty("Error Message", "Complete the 2fa process before accessing this application");
        return err;
    }

    private JsonElement gated(java.util.function.Supplier<JsonElement> op) {
        if (!configuration.isLoggedIn()) return notLoggedInError();
        try { return op.get(); }
        catch (Exception ex) {
            JsonObject err = new JsonObject();
            err.addProperty("Error", ex.getMessage() == null ? ex.toString() : ex.getMessage());
            return err;
        }
    }

    // ---------- Auth ----------
    public JsonElement totpLogin(String mobileNumber, String ucc, String totp) {
        return loginApi.totpLogin(mobileNumber, ucc, totp);
    }
    public JsonElement totpValidate(String mpin) { return loginApi.totpValidate(mpin); }

    public JsonElement logout() {
        if (!configuration.isLoggedIn()) return notLoggedInError();
        configuration.bearerToken = null;
        configuration.editSid = null;
        configuration.editToken = null;
        JsonObject ok = new JsonObject();
        ok.addProperty("State", "OK");
        ok.addProperty("message", "You have been successfully logged out");
        return ok;
    }

    // ---------- Orders ----------
    public JsonElement placeOrder(OrderApi.PlaceOrderRequest req) { return gated(() -> orderApi.placeOrder(req)); }
    public JsonElement modifyOrder(OrderApi.ModifyOrderRequest req) { return gated(() -> orderApi.modifyOrder(req)); }
    public JsonElement cancelOrder(String orderId, String amo, boolean verify) {
        return gated(() -> orderApi.cancelOrder(orderId, amo == null ? "NO" : amo, verify));
    }
    public JsonElement cancelCoverOrder(String orderId, String amo, boolean verify) {
        return gated(() -> orderApi.cancelCoverOrder(orderId, amo == null ? "NO" : amo, verify));
    }
    public JsonElement cancelBracketOrder(String orderId, String amo, boolean verify) {
        return gated(() -> orderApi.cancelBracketOrder(orderId, amo == null ? "NO" : amo, verify));
    }

    // ---------- Reports ----------
    public JsonElement orderReport() { return gated(orderApi::orderReport); }
    public JsonElement orderHistory(String orderId) { return gated(() -> orderApi.orderHistory(orderId)); }
    public JsonElement tradeReport(String orderId) { return gated(() -> orderApi.tradeReport(orderId)); }

    // ---------- Portfolio ----------
    public JsonElement positions() { return gated(portfolioApi::positions); }
    public JsonElement holdings() { return gated(portfolioApi::holdings); }
    public JsonElement limits(String segment, String exchange, String product) {
        String s = segment == null ? "ALL" : segment;
        String e = exchange == null ? "ALL" : exchange;
        String p = product == null ? "ALL" : product;
        return gated(() -> portfolioApi.limits(s, e, p));
    }

    // ---------- Pricing ----------
    public JsonElement marginRequired(MarginApi.MarginRequiredRequest req) {
        return gated(() -> marginApi.marginRequired(req));
    }
    public JsonElement quotes(List<Instrument> instruments, String quoteType) {
        return quotesApi.quotes(instruments, quoteType);
    }

    // ---------- Scrip ----------
    public JsonElement scripMaster(String exchangeSegment) { return gated(() -> scripApi.scripMaster(exchangeSegment)); }
    public JsonElement searchScrip(String exchangeSegment, String symbol, String expiry,
                                   String optionType, String strikePrice) {
        return gated(() -> scripApi.searchScrip(exchangeSegment, symbol, expiry, optionType, strikePrice));
    }

    // ---------- Streaming ----------
    public void subscribe(List<Instrument> instruments, boolean isIndex, boolean isDepth) {
        if (!configuration.isLoggedIn()) { if (onError != null) onError.accept(new IllegalStateException("not logged in")); return; }
        ensureSocket();
        try { neoWebSocket.getLiveFeed(instruments, isIndex, isDepth); }
        catch (Exception ex) { if (onError != null) onError.accept(ex); }
    }

    public void unSubscribe(List<Instrument> instruments, boolean isIndex, boolean isDepth) {
        if (!configuration.isLoggedIn()) { if (onError != null) onError.accept(new IllegalStateException("not logged in")); return; }
        ensureSocket();
        neoWebSocket.unSubscribeList(instruments, isIndex, isDepth);
    }

    public void subscribeToOrderFeed() {
        if (!configuration.isLoggedIn()) { if (onError != null) onError.accept(new IllegalStateException("not logged in")); return; }
        ensureSocket();
        try { neoWebSocket.getOrderFeed(); }
        catch (Exception ex) { if (onError != null) onError.accept(ex); }
    }

    private void ensureSocket() {
        if (neoWebSocket != null) return;
        neoWebSocket = new NeoWebSocket(
                configuration.editSid, configuration.editToken,
                configuration.serverId, configuration.dataCenter);
        neoWebSocket.onOpen = m -> { if (onOpen != null) onOpen.accept(m); };
        neoWebSocket.onMessage = m -> { if (onMessage != null) onMessage.accept(m); };
        neoWebSocket.onError = e -> { if (onError != null) onError.accept(e); };
        neoWebSocket.onClose = m -> { if (onClose != null) onClose.accept(m); };
    }
}

File: client/typescript-client/src/NeoAPI.ts
import { Configuration } from "./configuration";
import { RestClient, Response } from "./rest";
import { LoginApi } from "./api/loginApi";
import { OrderApi } from "./api/orderApi";
import { PortfolioApi } from "./api/portfolioApi";
import { MarginApi } from "./api/marginApi";
import { ScripApi } from "./api/scripApi";
import { QuotesApi } from "./api/quotesApi";
import { NeoWebSocket } from "./websocket/NeoWebSocket";
import {
  PlaceOrderRequest,
  ModifyOrderRequest,
  MarginRequiredRequest,
  Instrument,
  QuoteInstrument,
  MessageCallback,
  ErrorCallback,
  VoidCallback,
} from "./types";

/**
 * NeoAPI — top-level SDK entry point. Mirrors the Python NeoAPI class.
 */
export class NeoAPI {
  readonly configuration: Configuration;
  private rest: RestClient;
  private loginApi: LoginApi;
  private orderApi: OrderApi;
  private portfolioApi: PortfolioApi;
  private marginApi: MarginApi;
  private scripApi: ScripApi;
  private quotesApi: QuotesApi;
  private neoWebSocket?: NeoWebSocket;

  onOpen?: VoidCallback;
  onMessage?: MessageCallback;
  onError?: ErrorCallback;
  onClose?: VoidCallback;

  constructor(opts: {
    environment?: string;
    accessToken?: string;
    neoFinKey?: string;
    consumerKey?: string;
  } = {}) {
    this.configuration = new Configuration(opts.environment || "uat");
    this.configuration.bearerToken = opts.accessToken;
    this.configuration.neoFinKey = opts.neoFinKey;
    this.configuration.consumerKey = opts.consumerKey;

    this.rest = new RestClient(this.configuration);
    this.loginApi = new LoginApi(this.rest, this.configuration);
    this.orderApi = new OrderApi(this.rest, this.configuration);
    this.portfolioApi = new PortfolioApi(this.rest, this.configuration);
    this.marginApi = new MarginApi(this.rest, this.configuration);
    this.scripApi = new ScripApi(this.rest, this.configuration);
    this.quotesApi = new QuotesApi(this.rest, this.configuration);
  }

  private requireLogin(): Response | null {
    if (!this.configuration.isLoggedIn()) {
      return { "Error Message": "Complete the 2fa process before accessing this application" };
    }
    return null;
  }

  // ---------- Auth ----------
  totpLogin(mobileNumber: string, ucc: string, totp: string): Promise<Response> {
    return this.loginApi.totpLogin(mobileNumber, ucc, totp);
  }
  totpValidate(mpin: string): Promise<Response> {
    return this.loginApi.totpValidate(mpin);
  }
  logout(): Response {
    const e = this.requireLogin();
    if (e) return e;
    this.configuration.bearerToken = undefined;
    this.configuration.editSid = undefined;
    this.configuration.editToken = undefined;
    return { State: "OK", message: "You have been successfully logged out" };
  }

  // ---------- Orders ----------
  async placeOrder(req: PlaceOrderRequest): Promise<Response> {
    const e = this.requireLogin();
    if (e) return e;
    try {
      return await this.orderApi.placeOrder(req);
    } catch (err) {
      return { Error: (err as Error).message };
    }
  }
  async modifyOrder(req: ModifyOrderRequest): Promise<Response> {
    const e = this.requireLogin();
    if (e) return e;
    try {
      return await this.orderApi.modifyOrder(req, () => this.orderApi.orderReport());
    } catch (err) {
      return { Error: (err as Error).message };
    }
  }
  async cancelOrder(orderId: string, amo = "NO", verify = false): Promise<Response> {
    const e = this.requireLogin();
    if (e) return e;
    try {
      return await this.orderApi.cancelOrder(orderId, amo, verify, () => this.orderApi.orderReport());
    } catch (err) {
      return { Error: (err as Error).message };
    }
  }
  async cancelCoverOrder(orderId: string, amo = "NO", verify = false): Promise<Response> {
    const e = this.requireLogin();
    if (e) return e;
    try {
      return await this.orderApi.cancelCoverOrder(orderId, amo, verify, () => this.orderApi.orderReport());
    } catch (err) {
      return { Error: (err as Error).message };
    }
  }
  async cancelBracketOrder(orderId: string, amo = "NO", verify = false): Promise<Response> {
    const e = this.requireLogin();
    if (e) return e;
    try {
      return await this.orderApi.cancelBracketOrder(orderId, amo, verify, () => this.orderApi.orderReport());
    } catch (err) {
      return { Error: (err as Error).message };
    }
  }

  // ---------- Reports ----------
  async orderReport(): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.orderApi.orderReport(); } catch (err) { return { Error: (err as Error).message }; }
  }
  async orderHistory(orderId: string): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.orderApi.orderHistory(orderId); } catch (err) { return { Error: (err as Error).message }; }
  }
  async tradeReport(orderId?: string): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.orderApi.tradeReport(orderId); } catch (err) { return { Error: (err as Error).message }; }
  }

  // ---------- Portfolio ----------
  async positions(): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.portfolioApi.positions(); } catch (err) { return { Error: (err as Error).message }; }
  }
  async holdings(): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.portfolioApi.holdings(); } catch (err) { return { Error: (err as Error).message }; }
  }
  async limits(segment = "ALL", exchange = "ALL", product = "ALL"): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.portfolioApi.limits(segment, exchange, product); } catch (err) { return { Error: (err as Error).message }; }
  }

  // ---------- Pricing ----------
  async marginRequired(req: MarginRequiredRequest): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.marginApi.marginRequired(req); } catch (err) { return { Error: (err as Error).message }; }
  }
  quotes(instruments: QuoteInstrument[], quoteType?: string): Promise<Response> {
    return this.quotesApi.quotes(instruments, quoteType);
  }

  // ---------- Scrip ----------
  async scripMaster(exchangeSegment?: string): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.scripApi.scripMaster(exchangeSegment); } catch (err) { return { Error: (err as Error).message }; }
  }
  async searchScrip(
    exchangeSegment: string,
    symbol = "",
    expiry?: string,
    optionType?: string,
    strikePrice?: string,
  ): Promise<Response> {
    const e = this.requireLogin(); if (e) return e;
    try { return await this.scripApi.searchScrip(exchangeSegment, symbol, expiry, optionType, strikePrice); } catch (err) { return { Error: (err as Error).message }; }
  }

  // ---------- Streaming ----------
  async subscribe(instruments: Instrument[], isIndex = false, isDepth = false): Promise<void> {
    const e = this.requireLogin(); if (e) { console.log(e); return; }
    this.ensureSocket();
    await this.neoWebSocket!.getLiveFeed(instruments, isIndex, isDepth);
  }
  async unSubscribe(instruments: Instrument[], isIndex = false, isDepth = false): Promise<void> {
    const e = this.requireLogin(); if (e) { console.log(e); return; }
    this.ensureSocket();
    await this.neoWebSocket!.unSubscribeList(instruments, isIndex, isDepth);
  }
  async subscribeToOrderFeed(): Promise<void> {
    const e = this.requireLogin(); if (e) { console.log(e); return; }
    this.ensureSocket();
    await this.neoWebSocket!.getOrderFeed();
  }

  private ensureSocket(): void {
    if (this.neoWebSocket) return;
    this.neoWebSocket = new NeoWebSocket(
      this.configuration.editSid || "",
      this.configuration.editToken || "",
      this.configuration.serverId || "",
      this.configuration.dataCenter,
    );
    this.neoWebSocket.onOpen = (m) => this.onOpen?.(m);
    this.neoWebSocket.onMessage = (m) => this.onMessage?.(m);
    this.neoWebSocket.onError = (err) => this.onError?.(err);
    this.neoWebSocket.onClose = (m) => this.onClose?.(m);
  }
}
